essentials/src/main/java/org/dmfs/jems/function/composite/Chained.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2017 dmfs GmbH
+ *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.dmfs.jems.function.composite;
+
+import org.dmfs.jems.function.Function;
+
+
+/**
+ * A {@link Function} composed by chaining two other {@link Function}s like {@code y = outer(inner(x))}, i.e. the result of this {@link Function} is the result
+ * of the outer {@link Function} at the result of the inner {@link Function}.
+ *
+ * @author Marten Gajda
+ */
+public final class Chained<Argument, Intermediate, Result> implements Function<Argument, Result>
```

Comment:
Doesn't it start to go towards functional programming? It's like higher order functions.

I am not sure how this direction fits into the declarative OOP style, but as I see whenever we use the general `Function` it's starting to go towards functional. Because those are not _objects_ obviously, not things that we model and represent. But don't get me wrong, I am not against them. I am actually looking forward to use lambdas. And there are the obvious mapping, filtering use-case where passing in a sort of visitor function is clearly beneficial. 

In other words, I think everything can be expressed with functions as well, the question is do we want it, do we want to mix the two paradigms?
With these kind of classes, wouldn't we work towards using Java and Objects to work in a more functional paradigm?
Not sure if that's possible, whether other's have done it, what are the pitfalls, worth it or not, or whether it's better to change to other language in that case, like Kotlin.

A little bit related the question of `Function' or `Filter` in `Conditional`. It's true that `Function` is more general but does that mean we should use `Function` in place of all interfaces with an argument and return value? (I just mention it as kind of analogy, not asking for `Filter` in `Conditional`'.)

My feeling is that if the goal is to keep with declarative OOP, `Function` usage should be kept to minimal.

These are just my general thoughts atm, not sure if they add anything.